### PR TITLE
fix: Handle compression level correctly for different algorithms

### PR DESCRIPTION
**Pull request by srebhan**
- [ ] Updated associated README.md.
- [x] Wrote appropriate unit tests.
- [x] Pull request title or commits are in [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/#summary)

replaces #13424

This PR maps the given, standardized compression level to the individual values for each compression algorithm. Furthermore, we add a function to map a string configuration value to the defined levels.

### CODE REVIEW

**Comment by zekena2** [comment]
I still see it restrictive to have these 4 levels named to follow only the deflate packages. For example klauspost/zstd has only 4 levels supported out of the 22  and all of the 4 don't have 'None' in them so all options will compress another package for [lz4](https://github.com/pierrec/lz4) is using 9 levels, snappy doesn't even have a level. Expanding this list will be hard as well. Let's imagine we added snappy and the user wanted to use it thinking that it supports all these levels that we listed here then figuring out that it actually doesn't. Why shall we cause a confusion??

**Comment by srebhan** [comment]
> I see it restrictive to have these 4 levels named to follow only the deflate packages.  For example klauspost/zstd has only 4 levels supported out of the 22 and all of the 4 don't have 'None' in them so all options will compress another package for [lz4](https://github.com/pierrec/lz4) is using 9 levels [...]

I used those four as an example. If you like we can also convert this to have 9 levels. All I want is a standardized set of values because we cannot ask a user to read through **each package** to find out what to type into the config file. Furthermore, how often do you need the fine-grained 22-level setting **in practice**? If we _really_ need those (I really doubt) we can still add another option e.g. `WithNumericCompressionLevel`...

> snappy doesn't even have a level. Expanding this list will be hard as well. Let's imagine we added snappy and the user wanted to use it thinking that it supports all these levels that we listed here then figuring out that it actually doesn't. Why shall we cause a confusion??

Which confusion? In the `NewSnappyEncoder` you simply error out saying `"no compression levels supported"` if the level is set by an option...

**Comment by zekena2** [comment]
While the numbers are chosen are package specific but they're inherently generic levels that are used in all the other implementation of the compression algorithm (ie 1 will always be faster than 9 but 9 will be smaller in size) and if someone doesn't care about these numbers he can ignore them if he does care he should know what they mean. These numbers are actually standards by themselves within the compression Algorithm.

`WithCompressionLevel` This function accepts CompressionLevel. When I use it in a plugin without knowing anything I expect it to accept all the 4 values (None, Default, BestSpeed, BestCompression) The confusion comes when it gives me an error when I use None with zstd.

After all it's my opinion but I can go with any implementation and workaround it so it shouldn't be a blocker.

**Comment by srebhan** [comment]
@zekena2 yeah I guess the nine levels are a quasi-standard. So what is the mapping of levels 1-9? Can we just pass that into the libs? As I said, the new type was just an example...

For algorithms that do not support certain levels, we can document this and issue an error...

**Comment by srebhan** [comment]
@zekena2 is this what you had in mind?

**Comment by srebhan** [comment]
@zekena2 the issue is if some plugin decides not to call the checking function but directly instantiates the encoder directly... Furthermore, my approach is more modular as each encoder is responsible for checking its inputs so you have to change things in one position if an encoder changes / is added instead of having a global function...

**Comment by srebhan** [comment]
@powersj can you please give it another check?

**Review by powersj — APPROVED** [review]

File: internal/content_coding.go [hunk 1]
```diff
@@ -225,6 +234,13 @@ func NewZlibEncoder(options ...EncodingOption) (*ZlibEncoder, error) {
 		o(&cfg)
 	}
 
+	switch cfg.level {
+	case zlib.NoCompression, zlib.DefaultCompression, zlib.BestSpeed, zlib.BestCompression:
+		// Do nothing as those are valid levels
+	default:
+		return nil, fmt.Errorf("invalid compression level, only 0, 1 and 9 are supported")
```
Comment (Hipska) [inline internal/content_coding.go:241]: Inconsistent use of `fmt.Errorf` vs `errors.New`.

**Pull request merged**